File: Algorithms/8. String to Integer (atoi)/kotlin/solution.kt
class Solution {
    fun myAtoi(str: String): Int {
        if (str.length == 0) {
            return 0
        }

        var result = 0
        var plus = 1
        var hasNum = false
        var hasSign = false
        var startParse = false

        for (i in str.indices) {  
            if (str[i] != ' ' && !startParse) {
                startParse = true
            }

            if (!startParse) {
                continue
            }

            if (str[i] == '+' && !hasNum) {
                if (hasSign) {
                    break
                }
                hasSign = true
                plus = 1
                continue
            }

            if (str[i] == '-' && !hasNum) {
                if (hasSign) {
                    break
                }
                hasSign = true
                plus = -1
                continue
            }

            if (str[i] in '0'..'9') {
                hasNum = true

                val digit = str[i] - '0'                
                if (result > Int.MAX_VALUE/10 || 
                    (Int.MAX_VALUE/10 == result && Int.MAX_VALUE %10 < digit)) {
                    return if (plus == 1) Int.MAX_VALUE else Int.MIN_VALUE
                }

                result = result * 10 + digit
            } else {
                break
            }
        }

        result *= plus
        return result
    }
}

File: Algorithms/8. String to Integer (atoi)/swift/solution.swift
class Solution {
    func myAtoi(_ s: String) -> Int {
        var sign = 1
        var result = 0
        var i = 0
        while i < s.count && s[s.index(s.startIndex, offsetBy: i)] == " " {
            i += 1
        }
        if i < s.count && (s[s.index(s.startIndex, offsetBy: i)] == "-" || s[s.index(s.startIndex, offsetBy: i)] == "+") {
            sign = s[s.index(s.startIndex, offsetBy: i)] == "-" ? -1 : 1
            i += 1
        }
        while i < s.count && s[s.index(s.startIndex, offsetBy: i)] >= "0" && s[s.index(s.startIndex, offsetBy: i)] <= "9" {
            result = result * 10 + Int(String(s[s.index(s.startIndex, offsetBy: i)]))!
            i += 1
        }
        return sign * result
    }
}
